### PR TITLE
Update: use meta.docs.url in HTML formatter

### PR DESCRIPTION
**Pull request by pmcelhaney**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain:

In the HTML formatter, use the documentation URL from `meta.docs.url` instead of assuming every rule is a core rule. 

**What changes did you make? (Give an overview)**

- Changed the HTML formatter template so that "http://eslint.org/docs/rules/" is no longer hard-coded.
- Instead, look up the URL from the rule's meta (`meta.doc.url`)
- Pass the list of rules from cli.js to the formatter to make the above possible
- Update unit tests
- Add a check in the unit test that the URL is correct

**Is there anything you'd like reviewers to focus on?**

There may be a better way to mock the rules in the unit test.

### CODE REVIEW

File: lib/formatters/html.js [hunk 1]
```diff
@@ -82,30 +83,32 @@ function renderMessages(messages, parentIndex) {
             severityNumber: message.severity,
             severityName: message.severity === 1 ? "Warning" : "Error",
             message: message.message,
-            ruleId: message.ruleId
+            ruleId: message.ruleId,
+            url: rules.get(message.ruleId).meta.docs.url
```
Comment (j-f1) [inline lib/formatters/html.js]: Can you check this to ensure that `meta.docs` and `meta.docs.url` are not null?

**Comment by not-an-aardvark** [comment]
Thanks for the PR!

This is a core change, since it proposes adding an argument to formatters. Would you mind creating an issue for it? As described [here](https://eslint.org/docs/developer-guide/contributing/pull-requests#getting-started), we require an issue for all core changes.

Some initial thoughts:

* I like the idea here -- it seems like it would be useful to provide url info to formatters.
* If we do add a second argument to formatters, I think it should be an object (perhaps something like `{ rules: ruleMap }`) rather than the rule map itself. That way, if we want to provide additional information to formatters in a later enhancement, we could add it to the same object, so formatters wouldn't need to remember the argument order and do something like `function(results, unused, newInfo) {}`.
* `getRules` is currently somewhat slow, because it requires it requires loading all core rules from the filesystem, and copying a map of rules with ~250 elements. (Currently, we only load a core rule for the first time when it's actually used, which reduces startup time.) I wonder if it would be better to pass an object like `{ getRule: ruleName => rule }` to the formatter, so that it would only need to load the rules that are needed.
* As an open question, would it be better to just give formatters access to `rule.meta` or `rule.meta.docs` rather than giving access to a rule itself? Giving formatters access to a rule object itself would allow them to do things such as rerunning the rule, which seems like it could be confusing because formatters are just supposed to apply a transformation to the results. (Admittedly, I'm not sure why a formatter would want to rerun the rule, but ideally I think we should provide the minimal API needed to be useful.)

File: lib/formatters/html.js (hunk 1 above)
Comment (pmcelhaney) [reply lib/formatters/html.js]: Good catch. That would have broken on rules where `meta.docs` is not defined. (I'm assuming that's what you meant.)

**Comment by pmcelhaney** [comment]
If I remember correctly, we want to pass a copy of a rule’s `meta` to each reported problem in the `results` array. We can’t do that because it would bloat `results` when it’s serialized to JSON. 

If we keep the metadata out of the results object by passing the rules as a second argument, we cheat the JSON object out of valuable information. 

It now occurs to me that we could create a “v2” JSON formatter that serializes an object with two top level properties: `results` and `rules`. 

Would that work?

**Comment by nzakas** [comment]
Given that there hasn't been a consensus on the design, I'm closing this pull request. #9841 is where the conversation will continue.

**Pull request closed without merging**